Lazy Images: introduce deprecation notice in UI, and do not allow enabling it when disabled

jeherve opened the pull request:
Fixes #33166

## Proposed changes:

In this PR, we introduce:

- deprecation notices for the feature in the Jetpack plugin.
- logic to automatically disable the module when necessary.
- updated logic to detect whether the Lazy Images feature should be disabled. On top of the logic introduced in #33039, we now also check for future versions of WordPress (6.4) as well as sites running the development version of the Gutenberg plugin.

### Jetpack dashboard (Jetpack > Settings > Performance).

1. Always display a notice mentioning that the feature is no longer necessary, and explaining that it will be removed in November.
2. Do not allow enabling the module when it is disabled.
3. If you run a recent version of the Gutenberg plugin, customize the notice contents to explain that the feature is force-disabled now to avoid conflicts (see #33039).

<img width="1112" alt="Screenshot 2023-09-20 at 15 56 22" src="https://github.com/Automattic/jetpack/assets/426388/633ad974-38b2-4c42-ab0b-4c7f56b9eb63">
<img width="1087" alt="Screenshot 2023-09-20 at 15 56 08" src="https://github.com/Automattic/jetpack/assets/426388/35f57ec5-4b80-4538-b7d7-3efa1e380e47">
<img width="1126" alt="Screenshot 2023-09-20 at 15 52 31" src="https://github.com/Automattic/jetpack/assets/426388/23c511ca-a677-40d5-9503-79531cbc38de">

### Old Module list in the Jetpack plugin

Let's not make it available in the old module list **when it is disabled**. It does not need to be re-enabled.

### Jetpack Boost plugin

When the Lazy Images feature is flagged as conflicting with WordPress, we:

1. disable the feature
2. mark it as unavailable. It seems that right now, a feature that is marked as unavailable is just removed from the UI.

> **Warning**
> This means we'll need to make changes to the readme.txt file for the Boost plugin sooner rather than later. But I believe this will be tackled as part of #33170.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

- #33049
- pdWQjU-wV-p2

## Does this pull request change what data or activity we track or use?

* Yes, it introduces a new tracked link when one wants more information about the feature deprecation.

## Testing instructions:

**With the Jetpack plugin**

* Start on a site that **does not** run this branch.
* Go to Jetpack > Settings > Performance
* Enable Lazy Images
* Switch to this branch
* Notice the new notice in Jetpack > Settings > Performance.
* Enable the Gutenberg plugin
* Notice that the notice contents change.
* Notice that you cannot toggle the Lazy Images feature anymore.
* Load any post on the frontend.
* Back to Jetpack > Settings > Performance, the feature is now inactive.
* Go to Jetpack > Settings > Modules
* You should not see the Lazy Images module as available.

**With the Jetpack Boost plugin**

* Now enable the Boost plugin
* You should not see the Lazy Images functionality in the plugin UI.

You can also start with a brand new site with just the Boost plugin:

* Enable Boost on a site that runs WordPress 6.3 without Gutenberg.
* You should see the Lazy Images feature in Jetpack > Boost
* Now add the Gutenberg plugin
* The Lazy Images feature should now be gone.

jeherve committed:
Lazy Images: don't display module on old module list when disabled

Fixes #33166

Let's not make it available in the old module list when it is disabled. It does not need to be re-enabled.

jeherve committed:
Introduce new Gubenberg info state to use in React dashboard

jeherve committed:
Add new reducer to use Gutenberg data in React dashboard

jeherve committed:
Update UI to display the notice

jeherve committed:
Update module description

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/lazy-images-ui-deprecation

jeherve committed:
Remove unneeded backslash

jeherve committed:
Indicate upcoming deprecation in package readme

jeherve committed:
changelog

jeherve committed:
Bump versions

jeherve committed:
Break string up, it's easier to translate that way.

jeherve commented:
@kraftbj Before I open this up for reviews, I'd like to have your opinion on the overall UX of this change, as our temporary phase until the feature is fully removed.

Thank you!

Along this change, I would:

1. Add a JITM displayed to everyone with the module currently active. #33167
2. Update the support doc to explain the deprecation, since we're directing folks there (maybe I should direct folks to a specific anchor on the page, I thought about that, but maybe it's simpler to add a big notice at the top of the support doc) #33168
3. Update the Calypso UI to match this. #33169

kraftbj commented:
> maybe it's simpler to add a big notice at the top of the support doc)

Yeah, since this is a total deprecation and removal, no need to bury the notice in the text anywhere. Big and bold at the top is great.

Reviewing the UX itself now.

kraftbj commented:
Did not review the code so commenting, not reviewing.

Overall, I like this UX approach. It is pretty clean.

Things I noted:
* `wp jetpack module *` functions still list and let you activate lazy-images. I'm okay with this since it provides a way for folks to bring it back for whatever reason, but isn't something a regular person would stumble upon. Noting it for the sake of completeness.
* When using the Gutenberg force, the module itself is still active and displayed as such in the old modules list and the CLI. This isn't a big deal IMO, but would be nice to resolve.
* Bigger deal when using the Gutenberg force, since the module itself is still active, the Boost interface still shows Lazy Images as active. It won't load, of course, but that's confusing that the Jetpack regular interface and Boost interface show conflicting information.

Thus, my main suggestion would be:
* when using Gutenberg, let's on is_admin or within an admin hook, disable the module as well so we can have a consistent disabled experience for both states that would hit both plugins.

The lack of information in the Boost plugin is also a gap, but I know we're expecting HoG to handle the Boost-side of the deprecation.

jeherve committed:
Update logic in package to disable the Jetpack module

- When the site runs an incompatible version of WP (or Gutenberg), we deactivate the Jetpack module if it's active.
- When that is the case, we also do not load the feature at all.

jeherve committed:
Boost/Lazy Images: deactivate the matching Jetpack module

When necessary:

1. Deactivate the matching Jetpack module.
2. Make the feature not available in the UI.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/lazy-images-ui-deprecation

jeherve committed:
Remove logging

jeherve committed:
Remove unneeded changelog entry

jeherve committed:
Remove unneeded lock changelog entry

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/lazy-images-ui-deprecation

jeherve committed:
Update test to cover new use-cases

jeherve commented:
@kraftbj Thanks for taking a look! I took a pass at addressing your feedback:

- We now deactivate the Jetpack module from the Lazy Images package if we're in a conflicting environment.
- In the Jetpack Boost plugin, if we're in a conflicting environment ~we deactivate the Jetpack module if necessary and~ (not anymore, see https://github.com/Automattic/jetpack/pull/33208#discussion_r1333910076) we mark the Boost feature as unavailable. Boost modules include a `is_available` that seems to be good for that purpose, although when a module is not available it is completely removed from the UI. I don't think it's ideal, but I don't think that's something we should change in this PR. I think I'll need a review from the Heart of Gold team on those changes, since I want to be sure it's really okay to hook those added behaviors to `is_available`.

Let me know what you think of the UX now.

jeherve committed:
Fix comment

haqadn commented on projects/plugins/boost/app/modules/optimizations/lazy-images/class-lazy-images.php:
```diff
@@ -16,6 +16,13 @@ public static function get_slug() {
+		// In some scenarios, Lazy Images can conflict with other WordPress features.
+		// If we detect that, the feature should not be available.
+		if ( Jetpack_Lazy_Images::should_force_deactivate() ) {
+			Jetpack_Lazy_Images::disable();
```
My only recommendation is to avoid disabling the module here. This function is for checking the availability

haqadn commented on projects/plugins/boost/app/modules/optimizations/lazy-images/class-lazy-images.php:
```diff
@@ -16,6 +16,13 @@ public static function get_slug() {
+		// In some scenarios, Lazy Images can conflict with other WordPress features.
+		// If we detect that, the feature should not be available.
+		if ( Jetpack_Lazy_Images::should_force_deactivate() ) {
+			Jetpack_Lazy_Images::disable();
```
I would recommend that you avoid disabling the module here. This function is for checking the availability and disabling the module would be a side effect. Besides, it would disable the module in the Jetpack plugin.

In practice, returning false from this function if force deactivation is needed should be enough for it to work.

jeherve replied:
👍 Removed that in 72fc3ec3779fe0478ca373629a02a481a598e4c0

haqadn reviewed (commented):
Thanks for taking the time to update Boost. We will work on improving the UX and messaging on the Boost side so it doesn't feel like the feature is missing for no reason. However, marking this as unavailable is still suitable to avoid a conflict.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/lazy-images-ui-deprecation

jeherve committed:
Do not disable the Jetpack module from Jetpack Boost

See https://github.com/Automattic/jetpack/pull/33208#discussion_r1333910076

Co-authored-by: haqadn <haqadn@users.noreply.github.com>

anomiex commented on projects/packages/lazy-images/tests/php/test_class.lazy-images.php:
```diff
@@ -604,19 +604,58 @@ public function add_skip_lazy_class_to_attributes( $attributes ) {
+
+		$this->assertSame( $expected, Jetpack_Lazy_Images::should_force_deactivate() );
+
 		Constants::clear_constants();
+		$wp_version = $previous_version; // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
```
FYI, if the test fails these lines won't be reached to reset things. If you want to fix that, you could use a try-finally or move the resetting logic to a `tear_down()` method.

jeherve replied:
That's a good call. 9f62b74d690d452c7eee9f1cfba8f920c3395e99

anomiex commented on projects/packages/lazy-images/tests/php/test_class.lazy-images.php:
```diff
@@ -604,19 +604,58 @@ public function add_skip_lazy_class_to_attributes( $attributes ) {
+				),
+				true,
+			),
+		);
```
No test covering the `GUTENBERG_DEVELOPMENT_MODE` code path?

jeherve replied:
👍 Done in 4ae55d67024739a1d10087b720b546378e1ae973

anomiex commented on projects/plugins/jetpack/_inc/lib/admin-pages/class-jetpack-redux-state-helper.php:
```diff
@@ -241,6 +241,49 @@ public static function get_initial_state() {
+	 * @return array
+	 */
+	private static function get_gutenberg_initial_state() {
+		$initial_state = array(
```
This is only used once, it seems to me we may as well move it down to line 281. But feel free to ignore this suggestion if you disagree.

jeherve replied:
👍 52e06639eb7687885e90de51a1ca585ba9f01fbe

anomiex commented on projects/plugins/jetpack/changelog/update-lazy-images-ui-deprecation:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: compat
+
+Lazy Images: prepare feature for its deprecation, coming in November . TYou will be able to rely on Lazy loading features provided by WordPress itself.
```
```suggestion
Lazy Images: prepare feature for its deprecation, coming in November. You will be able to rely on Lazy loading features provided by WordPress itself.
```

jeherve committed:
Fix typo in changelog entry

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Simplify returns in method

See https://github.com/Automattic/jetpack/pull/33208#discussion_r1334488712

Co-authored-by: anomiex <anomiex@users.noreply.github.com>

jeherve committed:
Add test for development version of Gutenberg

See https://github.com/Automattic/jetpack/pull/33208#discussion_r1334484231

jeherve committed:
Update test to avoid messing with $wp_version when tests fail

See https://github.com/Automattic/jetpack/pull/33208/files#r1334483346

Co-authored-by: anomiex <anomiex@users.noreply.github.com>

anomiex commented:
Testing results:

> * Enable the Gutenberg plugin
> * Notice that the notice contents change.
> * Disable the Lazy Images feature

Can't, it's already showing as disabled.

> * Notice that you cannot enable it back.
> * Go to Jetpack > Settings > Modules
> * You should not see the Lazy Images module as available.

When I first try this, it shows the correct message and shows the slider as disabled but the option is not fully greyed out. The module is still present in the modules page and shows there as being *enabled*. After I reload a blog post, that apparently updates the database and now the option is fully greyed out and is no longer present on the modules page.

If I skip installing Gutenberg and just turn off the slider, it fully disables and greys out after the turning-off finishes.

> * You should see the Lazy Images feature in Jetpack > Boost

I note that this allows for turning the feature on when it's off, even though the corresponding slider in Jetpack's settings is disabled.

jeherve commented:
> Can't, it's already showing as disabled.

That's a good point. I need to update my testing instructions. Will do that now.

> When I first try this, it shows the correct message and shows the slider as disabled but the option is not fully greyed out. The module is still present in the modules page and shows there as being enabled. After I reload a blog post, that apparently updates the database and now the option is fully greyed out and is no longer present on the modules page.

Right, it will typically deactivate itself when Lazy Images is next loaded. I'm okay with this trade-off, personally.

> I note that this allows for turning the feature on when it's off, even though the corresponding slider in Jetpack's settings is disabled.

Yes, I think that's something that the folks in the Heart of Gold team will handle on their end, in #33170. I'll let them figure out the necessary UI updates that need to happen in the Boost plugin.

anomiex reviewed (approved):
Sounds good to me.

kraftbj committed:
Merge remote-tracking branch 'origin/trunk' into update/lazy-images-ui-deprecation
